### PR TITLE
Add sycl/gen backend and other sycl changes

### PR DESCRIPTION
**Pull request by uumesh**
- Adds the sycl/gen development
- fix CeedInt loop vars to CeedSize in sycl-ref to avoid integer overflow
- Required changes to compiler flags in Makefile

### CODE REVIEW

**Comment by jeremylt** [comment]
Note - as sycl is not in CI, manual running of the full test suite somewhere that has sycl is needed to approve this PR.

File: backends/sycl-gen/ceed-sycl-gen-operator.sycl.cpp
Comment (jrwrigh) [inline backends/sycl-gen/ceed-sycl-gen-operator.sycl.cpp:98]:
```suggestion
```
I'm getting "unused variable" warnings for `thread_2d`

Comment (uumesh) [reply backends/sycl-gen/ceed-sycl-gen-operator.sycl.cpp]: I have removed this variable in the following patch

File: backends/sycl-gen/ceed-sycl-gen-operator-build.sycl.cpp
Comment (jrwrigh) [inline backends/sycl-gen/ceed-sycl-gen-operator-build.sycl.cpp:675]:
```suggestion
               << ", P_out_" << i << ", Q_1D, r_tt_" << i << (dim > 1 ? ", s_B_out_" : "") << (dim > 1 ? std::to_string(i) : "") << ", s_G_out_" << i
```
trailing whitespace

File: Makefile
Comment (jrwrigh) [inline Makefile:469]:
```suggestion
# MAGMA Backends
```

ie. keep the original code.

File: backends/sycl-ref/ceed-sycl-ref-qfunction-load.hpp
Comment (jrwrigh) [inline backends/sycl-ref/ceed-sycl-ref-qfunction-load.hpp:11]:
```suggestion
CEED_INTERN int CeedQFunctionBuildKernel_Sycl(CeedQFunction qf);
```

We had a PR (https://github.com/CEED/libCEED/pull/1231) to make the function names here more consistent across backends. So I'm fairly certain we want to keep the original names.

File: backends/sycl-ref/ceed-sycl-ref-qfunction-load.sycl.cpp
Comment (jrwrigh) [inline backends/sycl-ref/ceed-sycl-ref-qfunction-load.sycl.cpp:29]:
```suggestion
extern "C" int CeedQFunctionBuildKernel_Sycl(CeedQFunction qf) {
```
Same here, keep original name.

File: backends/sycl-ref/ceed-sycl-ref-qfunction-load.sycl.cpp
Comment (jrwrigh) [inline backends/sycl-ref/ceed-sycl-ref-qfunction-load.sycl.cpp:166]:
```suggestion
  CeedCallBackend(CeedBuildModule_Sycl(ceed, code.str(), &impl->sycl_module));
  CeedCallBackend(CeedGetKernel_Sycl(ceed, impl->sycl_module, kernel_name, &impl->QFunction));
```
Keep original names

File: backends/sycl-ref/ceed-sycl-ref-qfunction.sycl.cpp
Comment (jrwrigh) [inline backends/sycl-ref/ceed-sycl-ref-qfunction.sycl.cpp:31]:
```suggestion
  if (!impl->QFunction) CeedCallBackend(CeedQFunctionBuildKernel_Sycl(qf));
```

File: backends/sycl-shared/ceed-sycl-shared-basis.sycl.cpp
Comment (jrwrigh) [inline backends/sycl-shared/ceed-sycl-shared-basis.sycl.cpp:249]: Revert to original names.

File: backends/sycl/ceed-sycl-common.hpp
Comment (jrwrigh) [inline backends/sycl/ceed-sycl-common.hpp:48]: Not sure if this change was intentional/necessary.

File: backends/sycl/ceed-sycl-compile.hpp
Comment (jrwrigh) [inline backends/sycl/ceed-sycl-compile.hpp:21]:
```suggestion
CEED_INTERN int CeedBuildModule_Sycl(Ceed ceed, const std::string &kernel_source, SyclModule_t **sycl_module,
                                        const std::map<std::string, CeedInt> &constants = {});
CEED_INTERN int CeedGetKernel_Sycl(Ceed ceed, const SyclModule_t *sycl_module, const std::string &kernel_name, sycl::kernel **sycl_kernel);
```

File: backends/sycl/ceed-sycl-compile.sycl.cpp
Comment (jrwrigh) [inline backends/sycl/ceed-sycl-compile.sycl.cpp:81]: I feel like the name of this function should be matched with the others (ie. `CeedLoadModule_Sycl`), but I'm not sure.

File: backends/sycl/ceed-sycl-compile.sycl.cpp
Comment (jrwrigh) [inline backends/sycl/ceed-sycl-compile.sycl.cpp:119]:
```suggestion
int CeedBuildModule_Sycl(Ceed ceed, const std::string &kernel_source, SyclModule_t **sycl_module,
```

File: backends/sycl/ceed-sycl-compile.sycl.cpp
Comment (jrwrigh) [inline backends/sycl/ceed-sycl-compile.sycl.cpp:143]:
```suggestion
int CeedGetKernel_Sycl(Ceed ceed, const SyclModule_t *sycl_module, const std::string &kernel_name, sycl::kernel **sycl_kernel) {
```

File: include/ceed/sycl.h
Comment (jrwrigh) [inline include/ceed/sycl.h:15]: I don't think we need this anymore? We moved to a generic `CeedSetStream`, so this doesn't need to be public facing anymore.

**Review by jrwrigh — COMMENTED** [review]
All tests pass on Sunspot, though for the `examples/petsc` tests I am seeing:
```
# Test:
# $ build/petsc-multigrid -ceed /gpu/sycl/ref -test -problem bp3 -degree 3 -simplex
ok 12 - PASS
# Test:
# $ build/petsc-bps -ceed /gpu/sycl/gen -test -problem bp3 -degree 3 -ksp_max_it_clip 50,50 -simplex
ok 14 - SKIP: not implemented petsc-bps /gpu/sycl/gen
# Test:
# $ build/petsc-multigrid -ceed /gpu/sycl/shared -test -problem bp3 -degree 3 -simplex
ok 13 - PASS
# Test:
# $ build/petsc-multigrid -ceed /gpu/sycl/gen -test -problem bp3 -degree 3 -simplex
ok 14 - SKIP: not implemented petsc-multigrid /gpu/sycl/gen
```

Not sure why these are being skipped, but the shared and ref backends are not. I'd imagine we'd want these tests to run for gen.

**Comment by jrwrigh** [comment]
Looks like the above skips were due to not supporting non-tensor elements. @uumesh How difficult would it be to add support for that given that it already exists in `/gpu/sycl/{ref,shared}`? If it's non-trivial, we could probably move that to another PR.

**Comment by jeremylt** [comment]
Adding non-tensor element support has been on the ToDo for years. Definitely not trivial

File: Makefile
Comment (nbeams) [reply Makefile:469]: Yeah, I agree with James. For anyone wondering, we actually do have multiple MAGMA backends even though we only build for one hardware type at a time -- `/gpu/*/magma` and `/gpu/*/magma/det` (deterministic version, uses ref restrictions).

File: Makefile
Comment (jrwrigh) [inline Makefile:278]:
```suggestion
sycl-gen.cpp   := $(sort $(wildcard backends/sycl-gen/*.sycl.cpp))
```

File: backends/sycl-ref/ceed-sycl-ref-qfunction-load.hpp
Comment (uumesh) [reply backends/sycl-ref/ceed-sycl-ref-qfunction-load.hpp:11]: Yes, I retained the function names in main for the most part. Not sure how I missed this.

Comment (jeremylt) [reply backends/sycl-ref/ceed-sycl-ref-qfunction-load.hpp]: Yeah, it looks like this PR missed updates in main. Thanks James for checking - this sort of consistency checking is tedious manual work but really helps maintainers when they need to go make a change across multiple backends

File: backends/sycl-gen/ceed-sycl-gen-operator-build.hpp
Comment (jeremylt) [inline backends/sycl-gen/ceed-sycl-gen-operator-build.hpp:12]: So here the updated convention would be 'CeedOperatorBuild_Sycl_gen', if I'm recalling correctly.

It's Ceed[Object][Action]_[Backend]

Comment (jeremylt) [reply backends/sycl-gen/ceed-sycl-gen-operator-build.hpp:12]: I think this specific function in CUDA/HIP was updated to CeedOperatorBuildKernel_Sycl_gen?

File: include/ceed/sycl.h
Comment (uumesh) [reply include/ceed/sycl.h:15]: @kris-rowe - It looks the registering this backend function is done inside the sycl-ref init function - https://github.com/argonne-lcf/libCEED/blob/fbf6e8e0622db949ec8e1942d099926308ba6a9d/backends/sycl-ref/ceed-sycl-ref.sycl.cpp#L41
Do you think this will take care of all backends or does it have to be included for each specific backend? Or maybe in the common sycl init function? Like here -  https://github.com/argonne-lcf/libCEED/blob/fbf6e8e0622db949ec8e1942d099926308ba6a9d/backends/sycl/ceed-sycl-common.sycl.cpp#L74

File: backends/sycl/ceed-sycl-compile.sycl.cpp
Comment (uumesh) [reply backends/sycl/ceed-sycl-compile.sycl.cpp:81]: Will modify the function name for this. But I hope using the 'Jit' tag in the function name for the others above are OK, since they are specific to the SYCL backend and provide a better sense of what is done in that function.

Comment (jrwrigh) [reply backends/sycl/ceed-sycl-compile.sycl.cpp:81]: Then perhaps it's not necessary to change this function name. I'm not as familiar with the other backends to know whether these functions have equivalents in CUDA or HIP. @jeremylt and/or @nbeams would know better.

Comment (jeremylt) [reply backends/sycl/ceed-sycl-compile.sycl.cpp]: CUDA and HIP also do JiT. We should use the equivalent function name from those backends but with a different suffix

Comment (jeremylt) [reply backends/sycl/ceed-sycl-compile.sycl.cpp]: Yeah, looks like the sporadic use of JiT in function names was removed from the other GPU backends

File: include/ceed/sycl.h
Comment (kris-rowe) [reply include/ceed/sycl.h]: Registering in sycl-ref init should be sufficient (e.g., like `GetPreferredMemType`), since the sycl-shared and sycl-gen backends will use the same function through delegation.

File: backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp
Comment (jeremylt) [inline backends/sycl-ref/ceed-sycl-ref-basis.sycl.cpp:659]: Whitespace changes like this were made for consistency as well. This sort of change in the sycl, sycl-ref, and sycl-shared folders should be undone

File: backends/sycl-shared/ceed-sycl-shared-basis.sycl.cpp
Comment (jeremylt) [inline backends/sycl-shared/ceed-sycl-shared-basis.sycl.cpp:263]: here

File: backends/sycl-shared/ceed-sycl-shared.sycl.cpp
Comment (jeremylt) [inline backends/sycl-shared/ceed-sycl-shared.sycl.cpp:51]: here

File: backends/sycl-shared/ceed-sycl-shared.sycl.cpp
Comment (jeremylt) [inline backends/sycl-shared/ceed-sycl-shared.sycl.cpp:67]: here

File: include/ceed/jit-source/sycl/sycl-ref-qfunction.h
Comment (jeremylt) [inline include/ceed/jit-source/sycl/sycl-ref-qfunction.h:23]: here

File: include/ceed/jit-source/sycl/sycl-shared-basis-tensor-templates.h
Comment (jeremylt) [inline include/ceed/jit-source/sycl/sycl-shared-basis-tensor-templates.h:28]: here

File: include/ceed/jit-source/sycl/sycl-shared-basis-tensor-templates.h
Comment (jeremylt) [inline include/ceed/jit-source/sycl/sycl-shared-basis-tensor-templates.h:179]: here

File: include/ceed/sycl.h
Comment (jeremylt) [inline include/ceed/sycl.h:11]: This file appears to be empty. Is it needed?

Comment (uumesh) [reply include/ceed/sycl.h:11]: File will be deleted in subsequent commit. It was intended to be kept as a placeholder to match other backends.

File: backends/sycl-ref/ceed-sycl-ref-qfunction-load.sycl.cpp
Comment (jeremylt) [inline backends/sycl-ref/ceed-sycl-ref-qfunction-load.sycl.cpp:175]: should be restored

File: backends/sycl-ref/ceed-sycl-ref-qfunction.sycl.cpp
Comment (jeremylt) [inline backends/sycl-ref/ceed-sycl-ref-qfunction.sycl.cpp:142]: should be restored

File: backends/sycl/ceed-sycl-common.sycl.cpp
Comment (jeremylt) [inline backends/sycl/ceed-sycl-common.sycl.cpp:25]:
```suggestion
```

**Review by jeremylt — COMMENTED** [review]
more consistency issues

File: backends/sycl/ceed-sycl-common.sycl.cpp
Comment (uumesh) [reply backends/sycl/ceed-sycl-common.sycl.cpp]: This function is removed in subsequent commit and function calls are modified to call the interface function

**Comment by jeremylt** [comment]
Looks good in terms of consistency now

**Review by jrwrigh — APPROVED** [review]
Reconfirmed that the branch passes all tests on Sunspot. So it's good to go from my end!

**Comment by jrwrigh** [comment]
@jeremylt Not sure if you want more time to look at before merging.

**Review by jeremylt — APPROVED** [review]

**Pull request merged**
